File: parse.rb
require 'json'

songs = []
id = 0
Dir.glob(__dir__+"/songs/*").each do |filename|
  id += 1
  file = File.new(filename, "r")
  song = {:name=>"", :melody=>"", :lyrics=>"", :id=>id}
  song[:name] = file.gets.strip
  song[:melody] = file.gets.strip
  while (line = file.gets)
    if line[0,1] == "{"
      meta = line
      line = file.gets
      while (line[0,1] != "}")
        meta += line
        line = file.gets
      end
      meta += line
      song[:meta] = JSON.parse(meta)
    else
      song[:lyrics] += line
    end
  end
  songs.push(song)
  file.close
end

File.open("songs.json","w") do |f|
  f.write(songs.to_json)
end

File: parse.mjs
import { promises, writeFile } from "fs";
import { join } from "path";
import fm, { test } from "front-matter";

const INPUT_DIRECTORY = "./songs/";
const OUTPUT_DIRECTORY = "./public/";
const OUTPUT_FILENAME = "songs.json";

async function parseSongFromFile(filename) {
  const data = await promises.readFile(
    join(INPUT_DIRECTORY, filename),
    "utf-8"
  );

  if (!test(data)) {
    throw Error(`File ${filename} is not in frontmatter format`);
  }

  const content = fm(data);
  const { title, melody, writers } = content.attributes;
  const lyrics = content.body;

  return { title, melody, writers, lyrics };
}

function writeSongs(filename, songs) {
  writeFile(filename, JSON.stringify(songs), function (err) {
    if (err) {
      console.log(err);
    } else {
      console.log("JSON saved to " + filename);
    }
  });
}

const filenames = await promises.readdir(INPUT_DIRECTORY);
const parsedSongs = await Promise.all(filenames.map(parseSongFromFile));
// Sort songs in alphabetical order
parsedSongs.sort((a, b) => a.title.localeCompare(b.title));
// Add id based on index
const songs = parsedSongs.map((song, i) => ({ id: i + 1, ...song }));

// If output directory does not exist, create it
try {
  await promises.mkdir(OUTPUT_DIRECTORY);
} catch (e) {}

writeSongs(join(OUTPUT_DIRECTORY, OUTPUT_FILENAME), songs);
